### ISSUE TITLE
getting NaNs for user vectors and item vectors when iterations is high (example 50,100) 

### ISSUE THREAD

**Issue by Ved-Piyush** [benfred/implicit#94]
Hi, 

I notice that with the deault number of iterations in implicit.als.AlternatingLeastSquares I do not get NaNs for the user and the item vecs, but if I increase the number of iterations to something like 50 or 100, sometimes there are just NaNs for the user and the item vecs. 

Any tips? 

Best,
Ved

**Comment by ita9naiwa** [comment]
This problems very hard to solve because the alternating least square model is to find exact solution iteratively.

Think the question ab = [[1,1]\1,1]] (2 by 2 matrix whose elements are all 1) (where a and b are variables that we want to find to fit the equation.)
Actually a is user factors and b is item factors and 2 is a rating or interaction matrix.

we first initialize b to be [0.001,0.001](we generally what we do when initialize the user,item factors) then the solution for a will be [1000,1000]

This has exact solution but, the important thing is  **ALS tends to make either user-factors or item-factors very large and another very small.** 

And as you know, what regularization does is make latent factors to be small enough.

Thus, the answer is use larger regularization parameter.

**Comment by Ved-Piyush** [comment]
Hi, 

Thanks for your answer, but this library doesn't implement the actual ALS algorithm, right? It implements the Conjugate Gradient version of it which find the user and item factors by still solving the ALS Cost function but in a Gradient Descent type of a method. Would that also converge quickly? Also, is there a way to gauge if the algorithm has converged or not?

**Comment by benfred** [comment]
@Ved-Piyush for checking convergence, there is a a calculate_training_loss parameter that will calculate and log a DEBUG message with the training loss to show if it is converging.

For the NaN values, do you have a example data that you can send me? Also can you verify that the confidence values in your input are all positive?

**Comment by FCHal** [comment]
Dear Ben,

First of all thanks for putting your methods online on github and your blog. They are really helpfull.

I am writing my thesis on this topic and I am having trouble with the 'slow' ALS method for implicit feedback.
I have a dataset with articles, users and their implicit feedback scores. These onfidence values in my input are all positive scores between 0 and 2. 

When I apply the method I get results up until around 2000 users (rows). With more than that I get arrays with NaN values only. Is the system overloaded? Or what causes the nan?

ita9naiwa suggested to increase the regularization term. But this did not help for me.
I have tried multiple regularization terms:  (000.1, 0.01, 0.1, 1, 2, 5, 10, 100), but none of them help.

Do you know what could be the problem? 
Would it be better to use the conjugate gradient descent? Or change the range of my confidence? (Something between 0-1)

Thanks in advance for your help.

**Comment by Ved-Piyush** [comment]
for alpha in alpha_g:
# # #
#
# # #
alpha_val = alpha
data_conf = sparse.csr_matrix((alpha_val*sparse_item_user))
data_conf.data +=1
data_conf = sparse.csr_matrix(data_conf.toarray().astype(np.float64))

the original AlS paper by Yehuda Koren defines

confidence of a user item count as 1 + alpha*r_ui

where r_ui is the number of times user u has used item i

so if you have to use different alpha values, you can see the code above

basically define a for loop with different values of alpha, multiply the alpha to every item in the dataset and then later add one to those multiplied values.

I ran into this problem all the time, what worked for me was to convert  the  dataset to the type np.float64. 

Give this a try.

**Comment by FCHal** [comment]
Dear Ved,

Thanks for your quick reply.
The code is not displaying well.

You suggest adding:

alpha_val = alpha
data_conf = sparse.csr_matrix((alpha_val*sparse_item_user))
data_conf.data +=1
data_conf = sparse.csr_matrix(data_conf.toarray().astype(np.float64))

in front of my code will work?

I will give it a try now.
Thanks very much

**Comment by FCHal** [comment]
My data conists of users and articles and my confidence is 'impl_feedback'.

Where do you suggest to put

for alpha in alpha_g:
  alpha_val = alpha
  data_conf = sparse.csr_matrix((alpha_val*sparse_item_user))
  data_conf.data +=1
  data_conf = sparse.csr_matrix(data_conf.toarray().astype(np.float64))



df_matrix_users['userID'] = df_matrix_users['userID'].astype("category")
df_matrix_users['article_id'] = df_matrix_users['article_id'].astype("category")


MF_csr0 = coo_matrix((df_matrix_users['impl_feedback'].astype(np.float64),    
                   (df_matrix_users['article_id'] .cat.codes,
                    df_matrix_users['userID'].cat.codes))).tocsr()

def nonzeros(m, row):
    """ returns the non zeroes of a row in csr_matrix """
    for index in range(m.indptr[row], m.indptr[row+1]):
        yield m.indices[index], m.data[index]

def alternating_least_squares(Cui, factors, regularization, iterations=20):
    users, items = Cui.shape

    X = np.random.rand(users, factors) * 0.01
    Y = np.random.rand(items, factors) * 0.01

    Ciu = Cui.T.tocsr()                                                            
    for iteration in range(iterations):
        least_squares(Cui, X, Y, regularization)
        least_squares(Ciu, Y, X, regularization)

    return X, Y

def least_squares(Cui, X, Y, regularization):
    users, factors = X.shape
    YtY = Y.T.dot(Y)

    for u in range(users):
        # accumulate YtCuY + regularization * I in A
        A = YtY + regularization * np.eye(factors)

        # accumulate YtCuPu in b
        b = np.zeros(factors)

        for i, confidence in nonzeros(Cui, u):
            factor = Y[i]
            A += (confidence - 1) * np.outer(factor, factor)
            b += confidence * factor

        # Xu = (YtCuY + regularization * I)^-1 (YtCuPu)
        X[u] = np.linalg.solve(A, b)

**Comment by Alex-Bujorianu** [comment]
I found out that the reason I was getting this error was because there were NaNs in the interaction matrix (which should have been 0s). The default hyperparams worked fine once that was fixed.
